src/GraphQL/Internal/AST.hs

```diff
@@ -45,24 +52,71 @@ module GraphQL.Internal.AST
 
 import Protolude hiding (Type)
 
-import Data.Char (isDigit)
+import qualified Data.Aeson as Aeson
 import qualified Data.Attoparsec.Text as A
-import GraphQL.Internal.Tokens (tok)
+import Data.Char (isDigit)
+import Test.QuickCheck (Arbitrary(..), elements, listOf)
 
--- TODO: Our GraphQL.Value.Name has more guarantees than AST.Name (which is
--- just a Text alias). We should change that so AST.Name is guaranteed valid.
+import GraphQL.Internal.Tokens (tok)
 
 -- * Name
 
-type Name = Text
+-- | A name in GraphQL.
+--
+-- https://facebook.github.io/graphql/#sec-Names
+newtype Name = Name { getNameText :: Text } deriving (Eq, Ord, Show)
```

Comment:
It's just record field syntax. It's pretty common for `newtypes` as a way of unwrapping the type.

Main point is composition, not point-free style. (e.g. `map getNameText names`). Encoder has a couple of motivating examples. Everything could be written with pattern matching, but we'd have to do much more typing.

Also note that if we wanted to use argument unpacking, we'd have to export a view pattern, because we don't want to export the constructor.  

Maybe renaming it to `unName` or `nameToText` would make it clearer? Similarly, I could write a whole separate function definition that does it.